### PR TITLE
[5.11.0] [U2] Fix the username mismatch issue when using x509 as the second step 

### PR DESCRIPTION
**Pull request by piyarathnalakmali**
## Purpose
Resolves [product-is #13381](https://github.com/wso2/product-is/issues/13381) by front-porting the following PRs

- #48 
- #49 
- #50 
- #51

Webapp changes are done in [wso2-support/identity-apps #208](https://github.com/wso2-support/identity-apps/pull/208)

### CODE REVIEW

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java [hunk 1]
```diff
@@ -522,6 +546,20 @@ private void validateUsingSubject(String subject, AuthenticationContext authenti
...
     }
 
+    /**
+     * Check the authenticated user's tenant domain and verify whether it from super tenant or different tenant.
+     *
+     * @param authenticatedUser Get the authenticated user object from the authentication context  .
+     */
+    private String getAuthenticatedUserName(AuthenticatedUser authenticatedUser) {
+
+        String userName = authenticatedUser.getAuthenticatedSubjectIdentifier();
```
Comment (somindatommy) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java:556]: Dont we need to a null check here?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateDataHolder.java [hunk 2]
```diff
@@ -64,4 +68,44 @@ public HttpService getHttpService() {
...
\ No newline at end of file
+
+    /**
+     * Set account lock service.
+     *
+     * @return account lock service
+     */
+    public AccountLockService getAccountLockService() {
+
+        return accountLockService;
```
Comment (somindatommy) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateDataHolder.java:82]: Lets throw an Runtime exception if this is null.  WDYT?

Comment (somindatommy) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateDataHolder.java:82]: Shall we do the same for other places as well.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java [hunk 3]
```diff
@@ -373,4 +379,91 @@ private static boolean isUserExists(String userName, AuthenticationContext authe
...
     }
 
-}
\ No newline at end of file
+    /**
+     * Check whether user account is locked or not.
+     *
+     * @param user Authenticated user.
+     * @return boolean account locked or not.
+     * @throws AccountLockServiceException
```
Comment (somindatommy) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java]: Complete the comment with exception description.

Comment (somindatommy) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java]: Shall we fix in other places as well?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java [hunk 4]
```diff
@@ -373,4 +379,91 @@ private static boolean isUserExists(String userName, AuthenticationContext authe
...
+    /**
+     * Get the user account state by checking  whether the user account is locked or not.
+     *
+     * @param subject userName that passed from the cert CN value.
+     * @return boolean account locked or not.
+     * @throws AccountLockServiceException
+     */
+    public static boolean isAccountLock(String subject) throws AccountLockServiceException {
+
+        //Get the tenantaware username & particular tenant domain
```
Comment (somindatommy) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java]: Formatting error

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/internal/X509CertificateServiceComponent.java [hunk 5]
```diff
@@ -99,4 +108,31 @@ protected void unsetHttpService(HttpService httpService) {
...
+
+        X509CertificateDataHolder.getInstance().setAccountLockService(null);
+    }
+
+    protected void setRealmService(RealmService realmService) {
+
+        if (log.isDebugEnabled()) {
+            log.debug("RealmService is set the X509 authenticator bundle.");
+        }
+
```
Comment (somindatommy) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/internal/X509CertificateServiceComponent.java]: Is this newline needed?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java (hunk 1 above)
Comment (piyarathnalakmali) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java:556]: null check is handled from the place where this method is being called.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java [hunk 6]
```diff
@@ -281,6 +285,25 @@ private void addOrValidateCertificate(String userName, AuthenticationContext aut
...
+                throw new AuthenticationFailedException("Account is locked for user: " + userName);
+            }
+
+            // Check whether user account is disable or not.
+            if (isAccountDisabled(getUsername(authenticationContext))) {
+                authenticationContext.setProperty(X509CertificateConstants.X509_CERTIFICATE_ERROR_CODE,
+                        X509CertificateConstants.USER_ACCOUNT_DISABLED);
+                throw new AuthenticationFailedException("Account is disabled for user: " + userName);
+            }
+        } catch (UserStoreException | AccountLockServiceException e) {
```
Comment (sahandilshan) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java:302]: Don't we need to throw the `e` with the exception?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateDataHolder.java (hunk 2 above)
Comment (piyarathnalakmali) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateDataHolder.java:82]: Fixed with ed2968b2c84d173790ff71b500a0268434981d01

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java (hunk 3 above)
Comment (piyarathnalakmali) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java]: Fixed with ed2968b2c84d173790ff71b500a0268434981d01

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java (hunk 4 above)
Comment (piyarathnalakmali) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java]: Fixed with ed2968b2c84d173790ff71b500a0268434981d01

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/internal/X509CertificateServiceComponent.java (hunk 5 above)
Comment (piyarathnalakmali) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/internal/X509CertificateServiceComponent.java]: Removed with ed2968b2c84d173790ff71b500a0268434981d01

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java (hunk 6 above)
Comment (piyarathnalakmali) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java:302]: Fixed with ed2968b2c84d173790ff71b500a0268434981d01

**Review by somindatommy — APPROVED** [review]

**Pull request merged**